aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -342,18 +343,49 @@ Tensor& eye_out_cpu(Tensor& result, int64_t n, int64_t m) {
 
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ full ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+namespace {
+
+// Performs dtype inference for full
+TensorOptions infer_full_options(
+  Scalar fill_value,
+  const TensorOptions& options) {
+
+  if (!options.has_dtype()) {
+    if (fill_value.isIntegral(true)) {
+      TORCH_WARN_ONCE(
+        "Deprecation warning: In PyTorch 1.6 torch.full with an integral ",
+        "fill_value will require a dtype or out argument. ",
+        "In PyTorch 1.7 an integral fill_value will return a LongTensor ",
+        "by default. ",
+        "Set the optional dtype or out arguments to suppress this warning. "
+      );
+    } else if (fill_value.isComplex()) {
+      auto scalar_type = (get_default_dtype() == ScalarType::Double) ?
+                            ScalarType::ComplexDouble :
+                            ScalarType::ComplexFloat;
+      return options.dtype(scalar_type);
+    }
+  }
+
+  return options;
+}
+
+} // anonymous namespace
+
 Tensor full(IntArrayRef size, Scalar fill_value, const TensorOptions& options) {
   if (options.layout() == kSparse) {
     AT_ERROR("full(...) is not implemented for sparse layout");
   }
-  auto result = at::empty(size, options);
+
+  auto result = at::empty(size, infer_full_options(fill_value, options));
   return result.fill_(fill_value);
 }
 
 Tensor& full_out(Tensor& result, IntArrayRef size, Scalar fill_value) {
```

Comment:
so...what should happen if the type given to full doesn't match the inference type?  It looks like arange just uses the result type, which seems reasonable.  What do you think?

Reply:
Was just having a similar conversation with @bhosmer. Luckily we don't need to think for ourselves since we can look at NumPy's behavior. In NumPy if the user specifies a dtype explicitly it wins. In full_like, the dtype of the tensor is used if a dtype isn't specified. Only as a last resort does NumPy infer the dtype from the fill value.

Reply:
oh nice, I forgot that NumPy actually implemented full_like.

I don't quite get the "Only as a last resort" sentence -- since this is _like, at least in PyTorch we always have a Tensor.  But I guess in NumPy since it can be array_like, it infers it based on the fill_value if it's not actually an ndarray?

Reply:
btw, do we have a test for this precedence behavior with full_like?

Reply:
I added a test for full_like. 

Good point that NumPy's behavior in the absence of the actual array is interesting. It must infer the dtype from the array's values in that case if no dtype is explicitly specified. But, as you mention, in PyTorch we don't have this issue.